segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -0,0 +1,186 @@
+                    txnData.setPersisted(true);
+                    return txnData;
+                }
+            }
+        } catch (IllegalStateException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new StorageMetadataException("Error while reading", e);
+        }
+
+        return TransactionData.builder()
+                .key(key)
+                .persisted(true)
+                .dbObject(TableKey.NOT_EXISTS)
+                .build();
+    }
+
+    /**
+     * Writes transaction data from a given list to the metadata store.
+     *
+     * @param dataList List of transaction data to write.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    protected void writeAll(Collection<TransactionData> dataList) throws StorageMetadataException {
+        ensureInitialized();
+        List<TableEntry> toUpdate = new ArrayList<>();
+        HashMap<TableEntry, TransactionData> entryToTxnDataMap = new HashMap<TableEntry, TransactionData>();
+        HashMap<TableKey, TransactionData> deletedKeyToTxnDataMap = new HashMap<TableKey, TransactionData>();
+        List<TableKey> keysToDelete = new ArrayList<>();
+        try {
+            for (TransactionData txnData : dataList) {
+                Preconditions.checkState(null != txnData.getDbObject());
+
+                long version = ((Long) txnData.getDbObject()).longValue();
+                if (null == txnData.getValue()) {
+                    val toDelete = TableKey.versioned(new ByteArraySegment(txnData.getKey().getBytes()),
+                            TableKey.NO_VERSION);
+                    keysToDelete.add(toDelete);
+                    deletedKeyToTxnDataMap.put(toDelete, txnData);
+                }
+
+                val arraySegment = serializer.serialize(txnData);
+
+                TableEntry tableEntry = TableEntry.versioned(
+                        new ByteArraySegment(txnData.getKey().getBytes()),
+                        arraySegment,
+                        version);
+                entryToTxnDataMap.put(tableEntry, txnData);
+                toUpdate.add(tableEntry);
+            }
+
+            // Now put uploaded keys.
+            List<Long> ret = this.tableStore.put(tableName, toUpdate, timeout).get();
+
+            // Update versions.
+            int i = 0;
+            for (TableEntry tableEntry : toUpdate) {
+                entryToTxnDataMap.get(tableEntry).setDbObject(ret.get(i));
+                i++;
+            }
+
+            // Delete deleted keys.
+            this.tableStore.remove(tableName, keysToDelete, timeout).get();
+            for (val deletedKey : keysToDelete) {
+                deletedKeyToTxnDataMap.get(deletedKey).setDbObject(TableKey.NOT_EXISTS);
+            }
+        } catch (RuntimeException e) {
+            throw e; // To make spotbugs happy.
+        } catch (java.util.concurrent.ExecutionException e) {
```

Comment:
Aren't these two catch blocks the same?

Reply:
You can use `Exceptions.unwrap` to help here.

Reply:
done